text/2021-02-24-coprocessor-plugin.md

```diff
@@ -0,0 +1,145 @@
+
+### API design
+
+To reduce the learning overhead, it'll be better that the API of the coprocessor plugin get closer to the client. Thus, it'll looks like the `RawClient` in the [Rust Client](https://github.com/tikv/client-rust) with extra txn-like methods e.g. `commit` and `lock`.
+
+```rust
+pub type Key = Vec<u8>;
+pub type Value = Vec<u8>;
+
+#[derive(Debug)]
+pub struct Region {
+    id: u64,
+    start_key: Key,
+    end_key: Key,
+    region_epoch: RegionEpoch,
+}
+
+#[derive(Debug)]
+pub struct RegionEpoch {
+    pub conf_ver: u64,
+    pub version: u64,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    KeyNotInRegion { key: Key, region: Region },
+    Deadlock { key: Key },
+    // More
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[async_trait]
+pub trait RawTransaction: Send {
+    /// Acquire memory lock for a key. All other trivial rpc requests or coprocessor
+    /// lock acquire to the same key will be blocked by this lock. This lock will be
+    /// released when the transaction is committed or dropped.
+    async fn lock(&self, key: Key) -> Result<()>;
+
+    async fn get(&self, key: Key) -> Result<Option<Value>>;
+    async fn scan(&self, key_range: Range<Key>) -> Vec<Value>;
+
+    async fn put(&mut self, key: Key, value: Value) -> Result<()>;
+    async fn delete(&mut self, key: Key) -> Result<()>;
+    async fn delete_range(&mut self, key_range: Range<Key>) -> Result<()>;
+
+    /// Returns when Raft message applied successfully.
+    async fn commit(self) -> Result<()>;
+}
+
+pub trait Coprocessor: Send + Sync {
+    fn on_raw_coprocessor_request(
+        &self,
+        region: Region,
+        request: Vec<u8>,
+        transaction: Box<RawTransaction>,
+    ) -> Result<Vec<u8>>;
+```
+
+## Future work
+
+### Official coprocessor
+
+Provide an official coprocessor plugin which defines common primitives like `bool`, `f32`, `UTF8String`, `Array<T>` along with codec for them, moreover, implements simple scanning expressions or even aggregations.
+
+Two reasons for that:
+
+1. To provide an example showing how to build a coprocessor plugin for TiKV and guaranteed to be always up-to-date.
+
+2. To comfort simple use cases to TiKV, so as to avoid forcing the users to develop a new plugin at the beginning.
+
+### Transaction mode
+
+Transaction mode is way more complicated than the RawKV mode. The current picture of txn plugin is an 'interactive' component with the client, which means, the client and the coprocessor will talk to each others during the request being processed. This is especially for resolving locks and coordinating the transaction among multiple coprocessors in multiple TiKV nodes.
+
+### Static plugin
+
+Discussed in the section above.
+
+### Migrate `tidb_query`
+
+When static plugin and txn plugin is implemented, we can move the coprocessor specific to TiDB out of TiKV, and the repository of TiDB might be a good new home.
```

Comment:
Have you evaluated if this is possible using the proposed framework?

Reply:
Not yet. This is the ultimate goal, but it will not block this RFC if this is not possible eventually. I think the most tricky part is the static plugin. Once we've figured it out, it will be quite trivial to do the migration.